src/replication.c

```diff
@@ -1401,36 +1405,35 @@ handle_snap_prepare_resp(replica_t *replica, mgmt_cmd_t *mgmt_cmd)
 }
 
 static int
-send_replica_snapshot(spec_t *spec, replica_t *replica, uint64_t io_seq,
-    char *snapname, zvol_op_code_t opcode, rcommon_mgmt_cmd_t *rcomm_mgmt)
+send_replica_volume_operation(spec_t *spec, replica_t *replica, uint64_t io_seq,
+		zvol_op_code_t opcode, size_t data_len, void *data, rcommon_mgmt_cmd_t *rcomm_mgmt)
 {
 	zvol_io_hdr_t *rmgmtio = NULL;
-	size_t data_len;
-	char *data;
 	zvol_op_code_t mgmt_opcode = opcode;
-	mgmt_cmd_t *mgmt_cmd;
+	mgmt_cmd_t *mgmt_cmd, *tmp_mgmt_cmd;
 	uint64_t num = 1;
-	int ret = 0;
+        int ret = 0;
 
 	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
-	memset(mgmt_cmd, 0, sizeof (mgmt_cmd_t));
+	memset(mgmt_cmd, 0, sizeof(mgmt_cmd_t));
 	mgmt_cmd->rcomm_mgmt = rcomm_mgmt;
-	data_len = strlen(spec->volname) + strlen(snapname) + 2;
 
 	BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
 
-	data = (char *)malloc(data_len);
-	snprintf(data, data_len, "%s@%s", spec->volname, snapname);
-
 	rmgmtio->io_seq = io_seq;
-
 	mgmt_cmd->io_hdr = rmgmtio;
 	mgmt_cmd->data = data;
 	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
 
 	MTX_LOCK(&replica->r_mtx);
-	//TODO: Add it as the second IO, rather than tailing
-	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+
+	// Adding as second IO in mgmt_cmd queue if queue contains items
+	tmp_mgmt_cmd = TAILQ_FIRST(&replica->mgmt_cmd_queue);
+	if (tmp_mgmt_cmd != NULL) {
```

Comment:
why we need this? If I am doing resize and snapshot one by one, is it possible it will be reversed means first we will take the snapshot then resize?

Reply:
Since we are taking lock on spec and also quiesce the other operation when other is going on (any time we will process either resize or snapshot). So there won't be any chance of having snapshots and resize in queue at given time. We are prioritizing snapshot/resize command than `ZVOL_OPCODE_REPLICA_STATUS and ZVOL_OPCODE_STATS`